### PR TITLE
Basic specification for key distribution for end-to-end cryptography.

### PR DESCRIPTION
**Pull request by NegativeMjark**
Basic specification for how a client uploads end-to-end keys to their home server.

### CODE REVIEW

File: specification/41_end_to_end_encryption.rst [hunk 1]
```diff
@@ -13,3 +13,283 @@ participating homeservers.
...
+    2) Alice requests Bob's public key and supported algorithms.
+
+      +----------------+  +------------+  +----------+
+      | Alice's Device |  | Alice's HS |  | Bob's HS |
+      +----------------+  +------------+  +----------+
+             |                  |               |
+             |=================>|==============>|
+               /keys/query        <federation>
+
+    3) Alice selects an algorithm claims any one-time keys needed.
```
Comment (ara4n) [inline specification/41_end_to_end_encryption.rst]: s/claims/and claims/

File: specification/41_end_to_end_encryption.rst [hunk 2]
```diff
@@ -13,3 +13,283 @@ participating homeservers.
...
+      "device_keys": {
+        "user_id": "<user_id>",
+        "device_id": "<device_id>",
+        "valid_after_ts": 1234567890123,
+        "valid_until_ts": 2345678901234,
+        "algorithms": [
+          "<algorithm-name>",
+        ],
+        "keys": {
+          "<algorithm>:<device_id>": "<key_base64>",
```
Comment (ara4n) [inline specification/41_end_to_end_encryption.rst]: Where are <algorithm-name> and <algorithm> etc BNF-style terms defined?  Is it <algorithm-name> or <algorithm_name>?

File: specification/41_end_to_end_encryption.rst [hunk 3]
```diff
@@ -13,3 +13,283 @@ participating homeservers.
...
+
+Encrypted messages are sent in the form.
+
+.. code:: json
+
+    {
+      "type": "m.room.encrypted",
+      "content": {
+        "algorithm": "<algorithm_name>"
+    } }
```
Comment (ara4n) [inline specification/41_end_to_end_encryption.rst]: shouldn't we show some placeholder fields here to show the algorithm-specific stuff?

File: specification/41_end_to_end_encryption.rst [hunk 4]
```diff
@@ -13,3 +13,283 @@ participating homeservers.
...
+
+Using Olm
+#########
+
+.. code:: json
+
+    {
+      "type": "m.room.encrypted",
+      "content": {
+        "algorithm": "m.olm.v1.curve25519-aes-sha2",
```
Comment (ara4n) [inline specification/41_end_to_end_encryption.rst]: folks on #matrix were asking for a breakdown of how the m.* namespace works for this, and how to interpret the implicit sub-fields here.  There were also objections from eternaleye(?) regarding the ambiguity of what precise flavours of AES and SHA2 we'd be referring to here.

File: specification/41_end_to_end_encryption.rst [hunk 5]
```diff
@@ -13,3 +13,283 @@ participating homeservers.
...
+        "sender_key": "<sender_curve25519_key>",
+        "ciphertext": {
+          "<device_curve25519_key>": {
+            "type": 0,
+            "body": "<base_64>"
+    } } } }
+
+The ciphertext is a mapping from device curve25519 key to an encypted payload
```
Comment (ara4n) [inline specification/41_end_to_end_encryption.rst]: s/encypted/encrypted/

File: specification/41_end_to_end_encryption.rst (hunk 4 above)
Comment (NegativeMjark) [inline specification/41_end_to_end_encryption.rst]: I've added a few notes under "Algorithms" to explain how the "m.*" namespace will work, and the rational for the algorithm naming.

**Comment by ara4n** [comment]
this LGTM other than still feeling that we're missing definitions of the BNF style terms.

File: specification/41_end_to_end_encryption.rst [hunk 6]
```diff
@@ -13,3 +13,326 @@ participating homeservers.
...
+
+There are two kinds of algorithms: messaging algorithms and key algorithms.
+Messaging algorithms are used to securely send messages between devices.
+Key algorithms are used for key agreement and digital signatures.
+
+Messaging Algorithm Names
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Messaging algorithm names use the extensible naming scheme used throughout this
+specification. Algorithm names that start with "m." are reserved for algorithms
```
Comment (Kegsay) [inline specification/41_end_to_end_encryption.rst]: Replace "m." with `` please.

File: specification/41_end_to_end_encryption.rst [hunk 7]
```diff
@@ -13,3 +13,326 @@ participating homeservers.
...
+The name "m.olm.v1.curve25519-aes-sha2" corresponds to version 1 of the Olm
+ratchet using Curve25519 for the initial key agreement, HKDF-SHA-256 for
+ratchet key derivation, Curve25519 for the DH ratchet, HMAC-SHA-256 for the
+hash ratchet, and HKDF-SHA-256, AES-256 in CBC mode, and 8 byte truncated
+HMAC-SHA-256 for authenticated encryption.
+
+Algorithm names should be short and meaningful. A name of "m.olm.v1" is too
+short. However a name of
+"m.olm.v1.ecdh-curve25519-hdkfsha256.hmacsha256.hkdfsha256-aes256-cbc-hmac64sha256"
+is too long despite giving a more precise description of the algorithm.
```
Comment (Kegsay) [inline specification/41_end_to_end_encryption.rst]: You should explain *why* it is bad to use long descriptive names like this.

File: specification/41_end_to_end_encryption.rst [hunk 8]
```diff
@@ -13,3 +13,326 @@ participating homeservers.
...
+        "<key_algorithm>": 50
+      }
+    }
+
+
+Downloading Keys
+~~~~~~~~~~~~~~~~
+
+Keys are downloaded as a collection of signed JSON objects. There
+will be a JSON object per device per user. If one of the user's
```
Comment (Kegsay) [inline specification/41_end_to_end_encryption.rst]: s/a/one/

File: specification/41_end_to_end_encryption.rst [hunk 9]
```diff
@@ -13,3 +13,326 @@ participating homeservers.
...
+
+.. code:: http
+
+    POST /keys/query HTTP/1.1
+    Content-Type: application/json
+
+    {
+      "device_keys": {
+        "<user_id>": ["<device_id>"]
+    } }
```
Comment (Kegsay) [inline specification/41_end_to_end_encryption.rst]: Wonky `}` matching

**Comment by Kegsay** [comment]
LGTM other than this - if conflicts are resolved and @ara4n is happy, we can merge this in.

**Comment by ara4n** [comment]
I think my & kegan's feedback still stands - I still think that a formal BNF-style description of the terms would be helpful, and kegan's corrections look relevant still.  Otherwise, it still LGTM.

**Pull request merged**